### PR TITLE
[Merged by Bors] - feat: the lambdaSyntax linter

### PR DESCRIPTION
**Pull request by grunweg**
The mathlib style guide states clearly that using \lambda for anonymous functions is disallowed, and the fun keyword is preferred. This linter enforces this.

---

I have decided to call the linter option `linter.style.lambdaSyntax`, fareshadowing the analogous renaming of options in #15971.
This linter is actually pretty simple: most of this PR are detailed tests.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mattrobball** [comment]
How about test with lambda used in other kinds of syntax nodes?

File: Mathlib/Tactic/Linter/Lint.lean [hunk 1]
```diff
@@ -211,6 +211,51 @@ initialize addLinter cdotLinter
...
+def lambdaSyntaxLinter : Linter where run := withSetOptionIn fun stx ↦ do
+    unless Linter.getLinterValue linter.lambdaSyntax (← getOptions) do
+      return
+    if (← MonadState.get).messages.hasErrors then
+      return
+    for s in findLambdaSyntax stx do
+      if let .atom _ "λ" := s.getArgs[0]! then
+        Linter.logLint linter.lambdaSyntax s m!"\
+        Please use 'fun' and not 'λ' to define anonymous functions.\n\
+        The latter syntax has been deprecated in mathlib 4."
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/Lint.lean:253]: Note that this approach does not allow:
```
example : ℕ → ℕ := set_option linter.lambdaSyntax false in λ _ ↦ 0
```
which would require walking the infotree instead.

Comment (grunweg) [reply Mathlib/Tactic/Linter/Lint.lean:253]: Good point! Not supporting this seems fine to me, but I have added a test to this effect.

**Comment by grunweg** [comment]
> How about test with lambda used in other kinds of syntax nodes?

I just added a lot of tests - which cannot be exhaustive; let me know if a particular kind of tests is missing!

**Comment by grunweg** [comment]
Rebased on master, and disabled this linter by default.

File: Mathlib/Tactic/Linter/Lint.lean [hunk 2]
```diff
@@ -247,6 +247,51 @@ initialize addLinter dollarSyntaxLinter
...
+namespace Style.lambdaSyntax
+
+/--
+`findLambdaSyntax stx` extracts from `stx` all syntax nodes of `kind` `Term.fun`. -/
+partial
+def findLambdaSyntax : Syntax → Array Syntax
+  | stx@(.node _ kind args) =>
+    let dargs := (args.map findLambdaSyntax).flatten
+    match kind with
+      | ``Parser.Term.fun => dargs.push stx
```
Comment (adomani) [inline Mathlib/Tactic/Linter/Lint.lean:275]:
```suggestion
      | ``Parser.Term.fun => dargs.push stx[0]
```
If from here you extract the `atom` directly...

File: Mathlib/Tactic/Linter/Lint.lean [hunk 3]
```diff
@@ -247,6 +247,51 @@ initialize addLinter dollarSyntaxLinter
...
+
+@[inherit_doc linter.style.lambdaSyntax]
+def lambdaSyntaxLinter : Linter where run := withSetOptionIn fun stx ↦ do
+    unless Linter.getLinterValue linter.style.lambdaSyntax (← getOptions) do
+      return
+    if (← MonadState.get).messages.hasErrors then
+      return
+    for s in findLambdaSyntax stx do
+      if let .atom _ "λ" := s.getArgs[0]! then
+        Linter.logLint linter.style.lambdaSyntax s m!"\
```
Comment (adomani) [inline Mathlib/Tactic/Linter/Lint.lean:287]:
```suggestion
      if let .atom _ "λ" := s then
        Linter.logLint linter.style.lambdaSyntax s m!"\
```
... then here you can simply log the warning on the `λ` more easily.

**Review by adomani — APPROVED** [review]
Just a very small stylistic suggestion: with the change, just `λ` should be underlined, with the current implementation, the whole `λ x => ...` gets underlined.

File: Mathlib/Tactic/Linter/Lint.lean [hunk 4]
```diff
@@ -247,6 +247,51 @@ initialize addLinter dollarSyntaxLinter
...
+# The `lambdaSyntax` linter
+
+The `lambdaSyntax` linter is a syntax linter that flags uses of the symbol `λ` to define anonymous
+functions, as opposed to the `fun` keyword. These are syntactically equivalent; mathlib style
+prefers the latter as it is considered more readable.
+-/
+
+/--
+The `lambdaSyntax` linter flags uses of the symbol `λ` to define anonymous functions.
+This is syntactically equivalent to the `fun` keyword; mathlib style prefers using the latter.-/
```
Comment (adomani) [inline Mathlib/Tactic/Linter/Lint.lean:260]:
```suggestion
This is syntactically equivalent to the `fun` keyword; mathlib style prefers using the latter.
-/
```

File: Mathlib/Tactic/Linter/Lint.lean [hunk 5]
```diff
@@ -247,6 +247,51 @@ initialize addLinter dollarSyntaxLinter
...
+def lambdaSyntaxLinter : Linter where run := withSetOptionIn fun stx ↦ do
+    unless Linter.getLinterValue linter.style.lambdaSyntax (← getOptions) do
+      return
+    if (← MonadState.get).messages.hasErrors then
+      return
+    for s in findLambdaSyntax stx do
+      if let .atom _ "λ" := s.getArgs[0]! then
+        Linter.logLint linter.style.lambdaSyntax s m!"\
+        Please use 'fun' and not 'λ' to define anonymous functions.\n\
+        The latter syntax has been deprecated in mathlib 4."
```
Comment (adomani) [inline Mathlib/Tactic/Linter/Lint.lean:289]:
```suggestion
        The 'λ'-syntax has been deprecated in mathlib4."
```
Just to be more explicit.

**Comment by adomani** [comment]
I just noticed that the changes in mathlib are `=>` vs `\mapsto`, though this is not what the linter enforces.  This should not necessarily change, but seems a little odd to modify code that the linter does not flag.

**Comment by grunweg** [comment]
Thanks for the suggestions, Damiano! I implemented all of them (tweaking the `stx` one slightly), and updated the tests.

**Review by jcommelin — COMMENTED** [review]
bors d+

**Comment by grunweg** [comment]
bors merge

**Pull request closed without merging**